### ISSUE TITLE
Fix vanishing gradient problem in attention mechanism and recurrent layer

### ISSUE THREAD

**Issue by barkincavdaroglu** [barkincavdaroglu/Link-Prediction-Mesh-Network#6]
Gradient becomes 0 in backpropagation for attention mechanism.

<img width="1460" alt="Screenshot 2022-12-29 at 19 15 22" src="https://user-images.githubusercontent.com/38387003/209980244-a0d6f308-ccda-4a80-8c88-6087c1cd602d.png">

Consider:

- [x] torch.clap values while calculating softmax of attention coefficients
- [x] Negative slope of LeakyReLU
- [x] Initializer for W and a parameters
- [x] Normalization

Looks like the problem originates from the LSTM/GRU layer. Gradients of this layer becomes zero very fast, and since the attention layer is behind the recurrent layer, gradients of attention layer are not properly updated.

<img width="500" alt="training_grad_2 0_norm_model lstm weight_hh_l0_epoch" src="https://user-images.githubusercontent.com/38387003/210097109-7bd010b2-cac4-4092-a552-fcdeb710e8c4.png">

Applying PowerTransformer to dataset and decreasing lr of RMSprop seems to have helped the gradients a little.
<img width="1513" alt="Screenshot 2022-12-31 at 14 28 32" src="https://user-images.githubusercontent.com/38387003/210135047-229cd368-3a77-4b73-b915-593ff6a2cad2.png">

Increasing num_epochs to 50 and setting gradient clip value to 5 has solved the issue.
<img width="1516" alt="Screenshot 2022-12-31 at 14 48 28" src="https://user-images.githubusercontent.com/38387003/210135637-970acbf9-bf82-481a-bedc-76e8a7e002e3.png">

Decreasing clip value could be further beneficial [will experiment during hyperparameter tuning].
